src/SAL/recovery/recovery_longhorn.c

```diff
@@ -0,0 +1,766 @@
+		}
+
+		total_size = cidstr_len + str_client_addr_len + 5 +
+			     cidstr_lenx_len;
+
+		/* hold both long form clientid and IP */
+		clientid->cid_recov_tag = gsh_malloc(total_size);
+
+		/* Can't overrun and shouldn't return EOVERFLOW or EINVAL */
+		(void) snprintf(clientid->cid_recov_tag, total_size,
+				"%s:%s",
+				cidstr_lenx, cidstr);
+	}
+
+	LogDebug(COMPONENT_CLIENTID, "Created client name [%s]",
+		 clientid->cid_recov_tag);
+}
+
+static int longhorn_recov_init(void)
+{
+	char host[NI_MAXHOST];
+	char payload[PAYLOAD_MAX];
+	char *response = NULL;
+	size_t response_size = 0;
+	char *version = NULL;
+	int err = 0;
+	int res = 0;
+
+	err = gethostname(host, sizeof(host));
+	if (err) {
+		LogEvent(COMPONENT_CLIENTID,
+				 "Failed to gethostname: %s (%d)",
+				 strerror(errno), errno);
+		return -errno;
+	}
+
+	LogEvent(COMPONENT_CLIENTID, "Initialize recovery backend '%s'", host);
+
+	version = generate_random_string(VERSION_BYTES);
+	assert(version != NULL);
+
+	memcpy(recov_version, version, VERSION_BYTES + 1);
+
+	snprintf(payload, sizeof(payload), "{\"hostname\": \"%s\", \"version\": \"%s\"}",
+		host, recov_version);
+
+	free(version);
+	version = NULL;
+
+	PTHREAD_RWLOCK_wrlock(&recov_lock);
+	res = http_call(HTTP_POST, LONGHORN_RECOVERY_BACKEND_URL,
+		payload, strlen(payload) + 1,
+		&response, &response_size);
+	PTHREAD_RWLOCK_unlock(&recov_lock);
+	if (res != 0) {
+		LogFatal(COMPONENT_CLIENTID, "HTTP call error: res=%d (%s)", res, response);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static void longhorn_recov_end_grace(void)
+{
+	char host[NI_MAXHOST];
+	char url[URL_MAX];
+	char payload[PAYLOAD_MAX];
+	char *response = NULL;
+	size_t response_size = 0;
+	int err = 0;
+	int res = 0;
+
+	err = gethostname(host, sizeof(host));
+	if (err) {
+		LogEvent(COMPONENT_CLIENTID,
+				 "Failed to gethostname: %s (%d)",
+				 strerror(errno), errno);
+		return;
+	}
+
+	LogEvent(COMPONENT_CLIENTID,
+			 "End grace for recovery backend '%s' version %s",
+			 host, recov_version);
+
+	snprintf(url, sizeof(url), "%s/%s", LONGHORN_RECOVERY_BACKEND_URL, host);
+	snprintf(payload, sizeof(payload), "{\"version\": \"%s\"}", recov_version);
+
+	PTHREAD_RWLOCK_wrlock(&recov_lock);
+	res = http_call(HTTP_PUT, url, payload, strlen(payload) + 1, &response, &response_size);
+	PTHREAD_RWLOCK_unlock(&recov_lock);
+	if (res != 0) {
+		LogFatal(COMPONENT_CLIENTID, "HTTP call error: res=%d (%s)", res, response);
+	}
+}
+
+static void longhorn_add_clid(nfs_client_id_t *clientid)
+{
+	char host[NI_MAXHOST];
+	char url[URL_MAX];
+	char payload[PAYLOAD_MAX];
+	char *response = NULL;
+	size_t response_size = 0;
+	char *encoded_cid_recov_tag = NULL;
+	int err = 0;
+	int res = 0;
+
+	err = gethostname(host, sizeof(host));
+	if (err) {
+		LogEvent(COMPONENT_CLIENTID,
+				 "Failed to gethostname: %s (%d)",
+				 strerror(errno), errno);
+		return;
+	}
+
+	longhorn_create_clid_name(clientid);
+	encoded_cid_recov_tag = url_encode(clientid->cid_recov_tag);
```

Comment:
I am curious why not just use `curl_easy_escape`? espeically you already use curl lib here.

Reply:
Ah, thanks for the hint. I was not aware of the curl API.

Reply:
Replaced with `curl_easy_escape()`. Thank you.